qiskit/backends/ibmq/ibmqjob.py

```diff
@@ -105,50 +102,79 @@ def cancel(self):
-        job_result = self._api.get_job(self._job_id)
-        stats = {'job_id': self._job_id}
-        self._status = None
-        _status_msg = None
-        if 'status' not in job_result:
-            self._exception = QISKitError("get_job didn't return status: %s" %
-                                          (pprint.pformat(job_result)))
-            raise QISKitError("get_job didn't return status: %s" %
-                              (pprint.pformat(job_result)))
-        elif job_result['status'] == 'RUNNING':
+        self._update_status()
+        stats = {
+            'job_id': self._job_id,
+            'status': self._status,
+            'status_msg': self._status_msg
+        }
+        if self._queue_position:
+            stats['queue_position'] = self._queue_position
+            # Reset once consumed to allow _update_status to regenerate the
+            # value if needed.
+            self._queue_position = None
+
+        return stats
+
+    def _update_status(self):
+        """Query the API to update the status."""
+        final_states = [
+            JobStatus.DONE,
+            JobStatus.CANCELLED,
+            JobStatus.ERROR
+        ]
+        if (self._status in final_states or
+                self._status == JobStatus.INITIALIZING):
+            return
+
+        try:
+            job_result = self._api.get_job(self._job_id)
+            if 'status' not in job_result:
+                raise QISKitError('get_job didn\'t return status: %s' %
+                                  pprint.pformat(job_result))
+
+        # pylint: disable=broad-except
+        except Exception as err:
+            self._status = JobStatus.ERROR
+            self._exception = err
+            self._status_msg = '{}'.format(err)
+            return
+
+        if job_result['status'] == 'RUNNING':
             self._status = JobStatus.RUNNING
-            # we may have some other information here
-            if 'infoQueue' in job_result:
-                if 'status' in job_result['infoQueue']:
-                    if job_result['infoQueue']['status'] == 'PENDING_IN_QUEUE':
-                        self._status = JobStatus.QUEUED
-                if 'position' in job_result['infoQueue']:
-                    stats['queue_position'] = job_result['infoQueue']['position']
+            queued, queue_position = self._is_job_queued(job_result)
+            if queued:
+                self._status = JobStatus.QUEUED
+            if queue_position:
+                self._queue_position = queue_position
+
         elif job_result['status'] == 'COMPLETED':
             self._status = JobStatus.DONE
+
         elif job_result['status'] == 'CANCELLED':
             self._status = JobStatus.CANCELLED
-        elif self.exception:
-            self._status = JobStatus.ERROR
-            if self._future_submit.exception():
-                self._exception = self._future_submit.exception()
-            self._status_msg = str(self.exception)
+
```

Comment:
How does this PR handle exceptions raised by `self._future_submit` which is done async and may occur later?

Reply:
There is an error in my assumptions regarding `_submit()` progressing the job to a final state. [This final assignment](https://github.com/QISKit/qiskit-sdk-py/pull/515/files/a105d5b92c20404a8f8e27bc3bbd027a5aa696d7#diff-1601b1931db3e408a0c7973d5605ff4aR313) happening at the end of the method invalidates de second paragraph of the explanation. I'll fix in the next commit. Thanks for the heads-up.